xla/service/gpu/fusions/triton/triton_fusion_emitter_large_test.cc

```diff
@@ -119,6 +119,50 @@ ENTRY e {
   EXPECT_TRUE(RunAndCompare(kHloText, ErrorSpec{/*aabs=*/1e-3, /*arel=*/1e-3}));
 }
 
+TEST_F(TritonGemmTest, MultiplePathsToSameOperandWorks) {
+  const std::string kHloText = R"(
+triton_computation {
+  p0 = f8e5m2[8192,512]{1,0} parameter(0)
+  p1 = f8e4m3fn[512,512]{1,0} parameter(1)
+  dot = bf16[8192,512]{1,0} dot(f8e5m2[8192,512]{1,0} p0, f8e4m3fn[512,512]{1,0} p1), lhs_contracting_dims={1}, rhs_contracting_dims={0}
+  p2 = f32[1]{0} parameter(2)
+  p3 = f32[1]{0} parameter(3)
+  p4 = pred[1]{0} parameter(4)
+  p5 = f32[1]{0} parameter(5)
+  p6 = pred[1]{0} parameter(6)
+  p7 = f32[1]{0} parameter(7)
+  select.0 = f32[1]{0} select(pred[1]{0} p4, f32[1]{0} p5, f32[1]{0} p3)
+  select.1 = f32[1]{0} select(pred[1]{0} p6, f32[1]{0} select.0, f32[1]{0} p3)
+  divide = f32[1]{0} divide(f32[1]{0} p7, f32[1]{0} select.1)
+  multiply.0 = f32[1]{0} multiply(f32[1]{0} p2, f32[1]{0} divide)
+  convert = bf16[1]{0} convert(f32[1]{0} multiply.0)
+  bitcast = bf16[] bitcast(bf16[1]{0} convert)
+  broadcast = bf16[8192,512]{1,0} broadcast(bf16[] bitcast), dimensions={}
+  multiply.1 = bf16[8192,512]{1,0} multiply(bf16[8192,512]{1,0} dot, bf16[8192,512]{1,0} broadcast)
+  ROOT bitcast.7436 = bf16[4,2048,4,128]{3,2,1,0} bitcast(bf16[8192,512]{1,0} multiply.1)
+}
+
+ENTRY e {
+  p0 = f8e5m2[8192,512]{1,0} parameter(0)
+  p1 = f8e4m3fn[512,512]{1,0} parameter(1)
+  p2 = f32[1]{0} parameter(2)
+  p3 = f32[1]{0} parameter(3)
+  p4 = pred[1]{0} parameter(4)
+  p5 = f32[1]{0} parameter(5)
+  p6 = pred[1]{0} parameter(6)
+  p7 = f32[1]{0} parameter(7)
+    ROOT fusion = bf16[4,2048,4,128]{3,2,1,0} fusion(p0,p1,p2,p3,p4,p5,p6,p7), kind=kCustom, calls=triton_computation, 
+  backend_config={"fusion_backend_config":
+      {"kind":"__triton_gemm", "triton_gemm_config":{"block_m":"64","block_n":"256","block_k":"32","split_k":"1","num_stages":"4","num_warps":"4","num_ctas":"1"}}}
+})";
+
+  MatchOptimizedHlo(kHloText, R"(
+; CHECK: triton
```

Comment:
So what happens with DFS? No fusion at all? A different, smaller, fusion?

Reply:
Note, the modified code does not construct fusions, it only traverses them to construct the correct order of emission. The problem is that the current implementation of this traversal does not allow multiple paths between two nodes. So far such graphs were never constructed, but they can be (that would be a separate change).